cmd/gardener-node-agent/app/app.go

```diff
@@ -288,6 +404,38 @@ func fetchAccessToken(ctx context.Context, log logr.Logger, restConfig *rest.Con
 	return nil
 }
 
+func fetchMachineName(fs afero.Afero) (string, error) {
+	machineNameByte, err := afero.ReadFile(fs, nodeagentv1alpha1.MachineNameFilePath)
+	if err != nil {
+		return "", fmt.Errorf("failed reading machine-name file: %w", err)
+	}
+	return strings.Split(string(machineNameByte), "\n")[0], nil
+}
+
+func fetchMachineNameFromNode(ctx context.Context, restConfig *rest.Config, fs afero.Afero, nodeName string) (string, error) {
+	if nodeName == "" {
+		return "", fmt.Errorf("node name is empty, cannot fetch machine name from node")
+	}
+	c, err := client.New(restConfig, client.Options{})
+	if err != nil {
+		return "", fmt.Errorf("unable to create client: %w", err)
+	}
+	node := &corev1.Node{}
+	if err := c.Get(ctx, client.ObjectKey{Name: nodeName}, node); err != nil {
+		return "", fmt.Errorf("unable to fetch node %q: %w", nodeName, err)
+	}
+	machineName, found := node.Labels[v1beta1constants.LabelMachineName]
+	if !found {
+		return "", fmt.Errorf("unable to get machine name. No %q label on node %q", v1beta1constants.LabelMachineName, node.Name)
+	}
+
+	if err := afero.WriteFile(fs, nodeagentv1alpha1.MachineNameFilePath, []byte(machineName), 0600); err != nil {
```

Comment:
We should only do this if the `nodeagentv1alpha1.MachineNameFilePath` file does not exist yet.

Reply:
`fetchMachineNameFromNode` function is only called when the file was not found ([ref](https://github.com/gardener/gardener/blob/84134e6a2f18f8cc6e34a7ca7ff3998b573766b7/cmd/gardener-node-agent/app/app.go#L273-L277)).